hips/hip-0019.md

```diff
@@ -0,0 +1,120 @@
+
+## Security Implications
+
+Integrating CUE introduces a new dependency into Helm, which must be carefully evaluated for security implications. Potential security considerations include:
+
+- **Code Injection Risks**: Ensuring that CUE file processing does not introduce vulnerabilities, such as code injection or execution of arbitrary code.
+
+- **Denial of Service**: Evaluating CUE processing performance to prevent potential denial-of-service attacks due to resource exhaustion.
+
+- **Dependency Management**: Regularly updating the CUE library to incorporate security patches and updates.
+
+A thorough security audit will be conducted to identify and mitigate any potential risks associated with CUE integration.
+
+## How to Teach This
+
+To educate users about this change:
+
+- **Documentation**: Update the Helm documentation to include comprehensive sections on using CUE, with examples, tutorials, and best practices.
+
+- **Migration Guides**: Provide detailed migration guides for users transitioning from JSON Schema to CUE, highlighting benefits and considerations.
+
+- **Workshops and Webinars**: Host online workshops and webinars to demonstrate the benefits of CUE and its effective use in Helm charts.
+
+- **Community Engagement**: Engage with the community through forums, GitHub discussions, and social media to answer questions, gather feedback, and provide support during the transition.
+
+- **Sample Charts**: Provide sample Helm charts utilizing CUE for validation, serving as references for users.
+
+## Reference Implementation
+
+A proof-of-concept implementation has been developed and is available in [PR #13236](https://github.com/helm/helm/pull/13236). This implementation demonstrates the integration of CUE with Helm and includes basic examples of values file validation using CUE.
+
+## Rejected Ideas
+
+- **Enhancing JSON Schema Support**: Extending the existing JSON Schema implementation to support newer drafts and features was considered but rejected due to the complexity and maintenance burden, as well as the limitations of JSON Schema in handling non-JSON formats like YAML.
+
+- **Developing a Custom Validation System**: Creating a custom validation system specifically for Helm was considered but deemed unnecessary given CUE's existing capabilities and ecosystem support.
+
+- **Using Alternative Tools**: Other configuration languages and validation tools were evaluated but did not offer the same level of compatibility, features, or community support as CUE.
```

Comment:
On this note, we have considered Pkl (https://pkl-lang.org/). I'm more familiar with CUE than Pkl. But Pkl would I believe provide both validation and rendering functionality.

That said, there is certainly rationale to keep value processing/validation and manifest rendering separate concerns. Which does mean we could choose separate technologies for each. ie. for values processing/validation we could support (JSON schema, CUE, Pkl), and for manifest rendering we could support (Pkl, gotemplate)

Overall, Helm I think needs to make a fairly strategic choice in how to approach these problem spaces. Personally, I would like to see better support in both. JSON schema (to the point of this HIP) is limited. And using unstructured text formatting ala. gotemplate for producing structured documents (Kubernetes object manifests) is a awkward at best solution.